### PR TITLE
MLFlow model logging code path improvement

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/11806?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11806/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11806
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Add `infer_code_path` param (default false) to automatically infer code paths. (In this PR, it only supports mlflow pyfunc flavor model logging, other flavors will support it in the follow-up PRs)

Design doc:
https://docs.google.com/document/d/144wAwgXsQ40C3dDsoObX0LfXp33aRbvEVCCq-VYgJqw/edit#heading=h.4cz970y1mk93

### How is this PR tested?

- [X] Existing unit/integration tests
- [X] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [x] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

For `mlflow.pyfunc.log_model` and `mlflow.pyfunc.save_model`, add infer_code_path param (default false) to automatically infer code paths. 

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/pyfunc/test_inferred_data_path.py
Comment (BenWilson2) [inline tests/pyfunc/test_inferred_data_path.py:61]: Can we update mod4.py with an import of some pypi package that would otherwise be included in the base pip requirements to illustrate to others looking at this test that we do not expect that import to be included due to it being external to the repo references?

File: mlflow/utils/model_utils.py
Comment (harupy) [inline mlflow/utils/model_utils.py:174]: This can be pulled out. We don't need to define this function every time `_infer_and_copy_code_paths` is called.

Comment (harupy) [reply mlflow/utils/model_utils.py:174]: Inner functions are error-prone.

File: mlflow/utils/model_utils.py
Comment (harupy) [inline mlflow/utils/model_utils.py:199]: I think `pathlib` can simplify this code.

File: mlflow/utils/model_utils.py
Comment (harupy) [inline mlflow/utils/model_utils.py:266]:
```suggestion
        sys.modules.pop(module, None)
```

just to make the code look safe.

File: mlflow/utils/requirements_utils.py
Comment (harupy) [inline mlflow/utils/requirements_utils.py:315]:
```suggestion
                           *record_full_module_args,
                        ],
```

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_data_path.py:11]: What is this line for?

Comment (harupy) [reply tests/pyfunc/test_inferred_data_path.py:11]: We can use https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.syspath_prepend

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_data_path.py:38]: we can use pathlib.

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_data_path.py:16]: `tmp_path` 
```suggestion
    return tmp_path / "model"
```

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_data_path.py:1]: /ai rewrite this code using `pathlib.`

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [reply tests/pyfunc/test_inferred_data_path.py:11]: We should avoid changing the global state like (`sys.path`) or environment variables. That can lead to weird and hard-to-debug errors. If you need to change it, make sure to clean up the change.

File: mlflow/utils/model_utils.py
Comment (harupy) [inline mlflow/utils/model_utils.py:221]: /ai remove "you" from this error message.

File: mlflow/pyfunc/__init__.py
Comment (harupy) [inline mlflow/pyfunc/__init__.py:2624]: can we run this before the first `mlflow_model.save`?

File: mlflow/pyfunc/__init__.py
Comment (harupy) [inline mlflow/pyfunc/__init__.py:2388]: let's update docstring

File: mlflow/utils/model_utils.py
Comment (harupy) [reply mlflow/utils/model_utils.py:221]: /ai I usually avoid using "we" or "you" in error messages. Guess why I do that.

File: mlflow/utils/model_utils.py
Comment (harupy) [inline mlflow/utils/model_utils.py:209]: same comment https://github.com/mlflow/mlflow/pull/11806/files#r1578932630

File: mlflow/utils/model_utils.py
Comment (harupy) [reply mlflow/utils/model_utils.py:221]: "you can't" sounds like we're blaming users.

File: mlflow/pyfunc/__init__.py
Comment (harupy) [inline mlflow/pyfunc/__init__.py:1]: can we add an examples in `examples/pyfunc` to demonstrate this feature?

File: tests/pyfunc/test_inferred_data_path.py
Comment (WeichenXu123) [reply tests/pyfunc/test_inferred_data_path.py:11]: btw, I want to run `tests/pyfunc/test_inferred_data_path.py` solely in CI, and use `tests/pyfunc` as its current working directory, :
`pushd pytest tests/pyfunc && pytest test_inferred_data_path.py && popd`

WDYT ?

Otherwise the "inferring code path" will capture all modules under `mlflow_repo`, which is annoying.
and we can set

File: mlflow/pyfunc/__init__.py
Comment (WeichenXu123) [reply mlflow/pyfunc/__init__.py:2624]: we can't, because when inferring, it has to create a child process and load the model (via `mlflow.pyfunc.load`) in the child process. Without the saved `MLmodel` file, loading model doen't work.

File: tests/pyfunc/test_inferred_data_path.py
Comment (WeichenXu123) [reply tests/pyfunc/test_inferred_data_path.py:61]: Sure. Good idea in testing.

File: .github/workflows/master.yml
Comment (harupy) [inline .github/workflows/master.yml:282]: can we use https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.chdir?

Comment (WeichenXu123) [reply .github/workflows/master.yml:282]: Good idea!

File: examples/pyfunc/train.py
Comment (harupy) [inline examples/pyfunc/train.py:57]: I'd create a separate example.

File: mlflow/utils/__init__.py
Comment (harupy) [inline mlflow/utils/__init__.py:302]:
```suggestion
def get_parent_module(module):
    return module[0: module.rindex(".")]
```

or `mod`.

File: mlflow/utils/docstring_utils.py
Comment (harupy) [inline mlflow/utils/docstring_utils.py:298]:
```suggestion
            include necessary python module files. Only python code files under current working directory
```

File: mlflow/utils/model_utils.py
Comment (harupy) [inline mlflow/utils/model_utils.py:201]: when you have a `pathlib.Path` object, you normally don't need to touch `os`.

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_data_path.py:93]: Can this class be simplified?

File: tests/pyfunc/test_inferred_data_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_data_path.py:40]:
```suggestion
def _walk_dir(path):
    return {str(p.relative_to(path)) for p in path.rglob("*") if p.is_file()}
```

Comment (harupy) [reply tests/pyfunc/test_inferred_data_path.py:40]: not sure if `_walk_dir` needs to be an iterator.

File: mlflow/utils/_capture_modules.py
Comment (harupy) [inline mlflow/utils/_capture_modules.py:70]: How much overhead does this add? An import statement is called many times. A small delay can sum up to a large delay. Not a blocker, but just curious.

File: mlflow/pyfunc/model.py
Comment (harupy) [inline mlflow/pyfunc/model.py:406]: Does `Model` class has this `code` attribute?

File: mlflow/utils/_capture_modules.py
Comment (WeichenXu123) [reply mlflow/utils/_capture_modules.py:70]: > An import statement is called many times

But because of python module caching, only the first importing execution triggers the hook.

File: tests/pyfunc/test_inferred_data_path.py
Comment (WeichenXu123) [reply tests/pyfunc/test_inferred_data_path.py:40]: nice simplification

File: mlflow/pyfunc/model.py
Comment (WeichenXu123) [reply mlflow/pyfunc/model.py:406]: Yes. the `MLmodel` yaml file has a field of `code`. So we need to set `code` attribute.

File: mlflow/utils/requirements_utils.py
Comment (harupy) [inline mlflow/utils/requirements_utils.py:348]:
```suggestion
*record_full_module_args,
```

File: mlflow/utils/_capture_modules.py
Comment (harupy) [inline mlflow/utils/_capture_modules.py:49]:
```suggestion
            if not self.record_full_module and is_absolute_import:
```

File: mlflow/pyfunc/model.py
Comment (harupy) [inline mlflow/pyfunc/model.py:46]: We have the same `FLAVOR_NAME` here:

https://github.com/mlflow/mlflow/blob/071adcf371b057d5172e43cfd20e020974a6682d/mlflow/pyfunc/__init__.py#L514

File: mlflow/pyfunc/model.py
Comment (harupy) [reply mlflow/pyfunc/model.py:406]:
```
% grep code mlflow/models/model.py
MODEL_CODE_PATH = "model_code_path"
        .. code-block:: python
        .. code-block:: python
        .. code-block:: python
        .. code-block:: python
    .. code-block:: python
    When logging model as code, this function can be used to set the model object
```

Looks like the Model class doesn't have a `code` attribute.

File: mlflow/utils/__init__.py
Comment (harupy) [reply mlflow/utils/__init__.py:302]:
```suggestion
def get_parent_module(_module):
    return _module[: _module.rindex(".")]
```

0 can be removed.

Comment (harupy) [reply mlflow/utils/__init__.py:302]: Any chance that `.` is not found?

File: mlflow/utils/_capture_modules.py
Comment (harupy) [reply mlflow/utils/_capture_modules.py:70]: Is it possible to measure the difference between `infer_code_paths=True` and `infer_code_paths=False`?

Comment (harupy) [reply mlflow/utils/_capture_modules.py:70]: or you can insert `print` to see how often it's called.

File: mlflow/utils/docstring_utils.py
Comment (harupy) [inline mlflow/utils/docstring_utils.py:303]: Any other scenarios where this doesn't work? If any, I think we should also clarify them.

Comment (harupy) [reply mlflow/utils/docstring_utils.py:303]: and this argument must be experimental. I feel like there are many unknown corner cases that the inference doesn't work.

File: mlflow/utils/__init__.py
Comment (harupy) [inline mlflow/utils/__init__.py:302]: /ai what happens if `.` is not found?

Comment (WeichenXu123) [reply mlflow/utils/__init__.py:302]: The `_CaptureImportedModules` implementation ensures logging full module name, and this is covered by `tests/pyfunc/test_inferred_data_path.py` (see file `tests/pyfunc/custom_model/mod1/__init__.py` )

File: mlflow/pyfunc/model.py
Comment (WeichenXu123) [reply mlflow/pyfunc/model.py:406]: You are right, 
we should use `mlflow_model.flavors[FLAVOR_NAME][CODE] = code_dir_subpath` instead of `mlflow_model.code = saved_code_subpath`

File: mlflow/utils/docstring_utils.py
Comment (WeichenXu123) [reply mlflow/utils/docstring_utils.py:303]: > Any other scenarios where infer_code_paths doesn't work or infer_code_paths should not be used? 

So far I can only think out the limitation of inferring current directory files.

File: mlflow/utils/_capture_modules.py
Comment (WeichenXu123) [reply mlflow/utils/_capture_modules.py:181]: oh, `mlflow.pyfunc.load_model(model_path)` internally invokes `importlib.import_module(pyfunc_conf[MAIN])`

File: mlflow/utils/docstring_utils.py
Comment (harupy) [reply mlflow/utils/docstring_utils.py:303]: what if your custom python script depends on  a json file?

Comment (WeichenXu123) [reply mlflow/utils/docstring_utils.py:303]: > what if your custom python script depends on a json file?

oh yes this is another case. let me update doc.

File: mlflow/utils/model_utils.py
Comment (WeichenXu123) [inline mlflow/utils/model_utils.py]: This causes strange error in `recipe` tests, need more investigation. I will make this change in followup PR.

Comment (harupy) [reply mlflow/utils/model_utils.py]: What is `strange error`?

File: mlflow/utils/_capture_modules.py
Comment (harupy) [reply mlflow/utils/_capture_modules.py:185]: I think that should be mentioned.

File: mlflow/utils/docstring_utils.py
Comment (harupy) [reply mlflow/utils/docstring_utils.py:303]: @WeichenXu123 It might not be a config.

File: mlflow/utils/docstring_utils.py
Comment (harupy) [inline mlflow/utils/docstring_utils.py:306]: I'd use `depends on` instead of `reads`.

File: mlflow/utils/model_utils.py
Comment (harupy) [reply mlflow/utils/model_utils.py]: This is import important, isn't it? Can't we really comment this out?

File: examples/pyfunc/infer_model_code_path.py
Comment (harupy) [inline examples/pyfunc/infer_model_code_path.py:16]: Do we need an sklearn model in this example?

File: examples/pyfunc/infer_model_code_path.py
Comment (harupy) [inline examples/pyfunc/infer_model_code_path.py:35]: do we need a child run to demonstrate the infer code path feature?

File: mlflow/utils/_capture_modules.py
Comment (harupy) [reply mlflow/utils/_capture_modules.py:74]: I think it's worth refactoring. The performance is important.

File: mlflow/utils/model_utils.py
Comment (WeichenXu123) [reply mlflow/utils/model_utils.py]: Let me repro the error in https://github.com/mlflow/mlflow/pull/11989 CI output
I can't repro it locally.

File: mlflow/utils/docstring_utils.py
Comment (WeichenXu123) [reply mlflow/utils/docstring_utils.py]: Yes I update doc saying it as resource files

File: mlflow/utils/docstring_utils.py
Comment (harupy) [inline mlflow/utils/docstring_utils.py:306]: `resource files (e.g. a config file)` sounds ambiguous. We should include non python files as clarity.

File: mlflow/utils/docstring_utils.py
Comment (harupy) [inline mlflow/utils/docstring_utils.py:306]:
```suggestion
    If your custom python module depends on non-python files (e.g. a JSON file) with a relative
```

File: examples/pyfunc/README.md
Comment (harupy) [inline examples/pyfunc/README.md:31]:
```suggestion
$ python infer_model_code_paths.py
```

nit

File: examples/pyfunc/infer_model_code_path.py
Comment (harupy) [inline examples/pyfunc/infer_model_code_path.py:1]: infer_model_code_paths.py -> infer_model_code_paths.py

File: tests/pyfunc/test_inferred_code_path.py
Comment (harupy) [inline tests/pyfunc/test_inferred_code_path.py:13]: Didn't `monkeypatch.chdir(path)` work?

Comment (WeichenXu123) [reply tests/pyfunc/test_inferred_code_path.py:13]: oh, we need both `monkeypatch.chdir(path)` and `sys.path.insert(0, ".")`.

the `.` by default is not in `sys.path`

Comment (harupy) [reply tests/pyfunc/test_inferred_code_path.py:13]: can we use https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.syspath_prepend then?

File: .github/workflows/master.yml
Comment (harupy) [inline .github/workflows/master.yml:280]: do we still need this?

**Review by harupy — APPROVED** [review]
LGTM once @BenWilson2 approves the docs.

File: .github/workflows/master.yml
Comment (WeichenXu123) [reply .github/workflows/master.yml:280]: Let me try.

File: mlflow/utils/docstring_utils.py
Comment (BenWilson2) [inline mlflow/utils/docstring_utils.py:299]:
```suggestion
            under current working directory are automatically inferrable. The default value is ``False``.
```

**Review by BenWilson2 — APPROVED** [review]
LGTM! I'll work on a quick PR to get https://mlflow.org/docs/latest/model/dependencies.html?highlight=code_paths updated with these changes

**Pull request merged**
